### PR TITLE
Add the ability to rolling upgrade cluster gracefully

### PR DESCRIPTION
**Pull request by baluchicken**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    |yes
| API breaks?     | no
| Deprecations?   | no
| License         | Apache 2.0

### What's in this PR?

This PR contains some refactor and the initial implementation of rolling upgrade

### Checklist

- [x] Implementation tested
- [] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [] User guide and development docs updated (if needed)

### CODE REVIEW

File: pkg/kafkautil/healthcheck.go [hunk 1]
```diff
@@ -38,3 +40,32 @@ func (k *kafkaClient) OfflineReplicaCount() (int, error) {
...
+			replicas, err := k.client.Replicas(topic, partition)
+			if err != nil {
+				return false, err
+			}
+			isrreplicas, err := k.client.InSyncReplicas(topic, partition)
+			if err != nil {
+				return false, err
+			}
+			if len(replicas) != len(isrreplicas) {
+				log.Info("not all replicas are in sync")
```
Comment (pepov) [inline pkg/kafkautil/healthcheck.go]: possibly worth to gather all the out of sync replicas and log which are out of sync? for debugging/troubleshooting...

File: pkg/resources/kafka/kafka.go [hunk 2]
```diff
@@ -245,7 +246,7 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 				r.pvc,
```
Comment (pepov) [inline pkg/resources/kafka/kafka.go]: is this inner for loop still required?

File: pkg/resources/kafka/kafka.go [hunk 3]
```diff
@@ -299,9 +300,9 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 			r.pod,
```
Comment (pepov) [inline pkg/resources/kafka/kafka.go]: is this for loop still required?

File: pkg/kafkautil/client.go [hunk 4]
```diff
@@ -73,13 +81,11 @@ func New(opts *KafkaConfig) (client KafkaClient, err error) {
 	}
 
 	if kclient.brokers, err = kclient.DescribeCluster(); err != nil {
-		kclient.admin.Close()
```
Comment (tinyzimmer) [inline pkg/kafkautil/client.go]: I think this is necessary still for garbage collection. If it got to the point where it is trying to `DescribeCluster`, it has already successfully opened an admin client. If that errs out then we want to make sure we close it I think.

File: api/v1alpha1/kafkacluster_types.go [hunk 5]
```diff
@@ -192,12 +191,12 @@ func (bConfig *BrokerConfig) GetResources() *corev1.ResourceRequirements {
 	}
 	return &corev1.ResourceRequirements{
 		Limits: corev1.ResourceList{
-			"cpu":    resource.MustParse("500m"),
-			"memory": resource.MustParse("1Gi"),
+			"cpu":    resource.MustParse("1500m"),
```
Comment (bonifaido) [inline api/v1alpha1/kafkacluster_types.go]: I would make these resource requirements configurable through the CR. People usually have different usage patterns.

File: pkg/kafkautil/healthcheck.go (hunk 1 above)
Comment (tinyzimmer) [reply pkg/kafkautil/healthcheck.go]: This is actually getting done by the kafkatopic CR itself too for whatever thats worth. At least if the topic was made through that. The status will be kept up to date and detail any out of sync replicas along with partition leaders and such.

File: api/v1alpha1/kafkacluster_types.go (hunk 5 above)
Comment (baluchicken) [reply api/v1alpha1/kafkacluster_types.go]: You are absolutely right, but we are planning to deal with this on a separate PR. This one is getting huge

Comment (bonifaido) [reply api/v1alpha1/kafkacluster_types.go]: sure thing

File: pkg/kafkautil/client.go (hunk 4 above)
Comment (baluchicken) [reply pkg/kafkautil/client.go]: Thanks, I just added this line

**Pull request merged**